Python/Product/Analysis/LanguageServer/ILanguageServerExtension.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using Microsoft.PythonTools.Projects;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public interface ILanguageServerExtension {
+        /// <summary>
+        /// Called when the extension is registered for a language server.
+        /// </summary>
+        void Register(Server server);
```

Comment:
Can we separate entry point from the extension itself?
```cs
public interface ILanguageServerExtensionProvider {
    Task<ILanguageServerExtension[]> GetExtensionsAsync(Server server, CancellationToken cancellationToken);
}
```

Also, `Server` seems too much for an input. Can we only provide an interface with required parameters?

Reply:
Or even better, a separate object instance.

Reply:
We'd basically end up with one provider per extension, so there's nothing gained here, and we still need to be able to pass in parameters. The IDE will likely have something like this though (e.g. in VS we use MEF to discover extensions, and then the provider gives us the info we need to pass over JSON to load the extension).

As for providing `Server`, I would prefer to pass `ServerBase`, but ultimately these extensions are for the Python Language Server and so probably won't ever be used with another one. Having the base class helps keep _us_ honest about "what is in LSP vs. what have we added", which is why it's there.

Reply:
+1 on the interface. Prevents extension from having recompile when class changes. Later we can put interfaces into a separate interop assembly. Also, allows extension to mock server for testing.

Reply:
It doesn't depend on the structure of `Server`, as there is no inheritance involved. And `Server` is already public API, which means we can't change existing public members anyway (adding new members is fine).

If we want to provide `ServerBase` as the parameter and make extensions downcast, that's fine by me. Given the need to pass `PythonAst` instances around, there isn't really a world where we can build extensions without depending on `Microsoft.PythonTools.Analysis.dll` anyway, so we may as well make it as easy as possible to build the extension.

Reply:
There are several reasons why extension instantiation should be separated from extension itself:
- Lifetime. Right now, it is kinda equal to the `Server`, but we may need to change this in the future (for instance, recreate extension if it throws too many exceptions), and extension writers may want to track instances without static collections that are always a problem for testing.
- Half-initialized object. It is always a question what can be done in ctor and what must be done in `Register`. Good example of how it ends is instantiation of tool windows and all the problems VS have with it right now. It is better to give extension writer freedom to return fully initialized object. This is also better for tests, cause instead of two calls from host test has to simulate only one.
- Read-only fields. Extension code will need far less null checks for fields that are assigned in ctor.
- Factory pattern and SRP

> We'd basically end up with one provider per extension

Is it a limitation of VSC?

> we may as well make it as easy as possible to build the extension.

That includes ability to test functionality without instantiation of heavy objects. That's why it is better to pass interface. You can take CPS with their 'SwitchToMainThreadAsync' that can't be mocked because of the struct awaiter. As a result, everyone create a wrapper, so all the benefits of that awaiter got swallowed by those wrappers.

`Server` is most long-living object which contains tons of public methods that aren't intended to be called from extension (including LSP handlers). Public object was designed for access from external host inside some third-party project. It is different from loading extension into our host and give it access inside Analysis workflow. Extension should have access to the parts that are designed for it and nothing more.

Reply:
>>We'd basically end up with one provider per extension

>Is it a limitation of VSC?

We have one `Server` instance per process. Each extension can only be loaded once per `Server` instance. Hence, each provider is only ever going to be instantiated once and used once to provide one instance of the extension. So we may as well merge them and have the factory be `.ctor`.

Reply:
`IServer` makes it easy to mock. Extension may want to have tests without actually running the process.

Reply:
What would `IServer` have that `ServerBase` doesn't have?

Reply:
Class cannot be mocked. Although it is possible to derive some sort of a test class from base, common practice is to write tests against interfaces using mocking frameworks like NSubstitute or TypeMoq. Also, interfaces eventually can go to a separate interop assembly and extension tests the won't need actual LS assembly.

IMO common practice is to expose interfaces unless class is trivial and is easy to create such as range/point/eventargs. Although I found that IRange is way more useful than Range/Span classes that VS and VSC use.

Reply:
Okay, adding an `IServer` class with the same interface as `ServerBase` so it can be mocked.

Reply:
I'll add a provider class as well. Shouldn't add too much code to every extension...

Reply:
I'm treating the interop assembly as out of scope right now though. When we move everything out, we'll just have to do type forwarding for them all